pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,244 @@
+---
+title: "Understanding JSON Schema Lexical and Dynamic Scopes"
+date: "2024-01-19"
+tags:
+  - Fundamentals
+  - Scopes
+type: Engineering
+cover: /img/posts/2024/understanding-lexical-dynamic-scopes/cover.webp
+authors:
+  - name: Juan Cruz Viotti
+    photo: /img/avatars/jviotti.webp
+    link: https://www.jviotti.com
+    byline: O'Reilly author, founder of Intelligence.AI and Sourcemeta
+excerpt: "A deep dive on lexical and dynamic scopes, essential for understanding more advanced topics like dynamic referencing"
+---
+
+Most of the keywords defined by the JSON Schema organization can be either
+evaluated on their own, or by only considering the values of their adjacent
+keywords. For example, the `type` keyword is independent of any other keywords,
+while the `minContains` and `maxContains` keywords depend on the `contains`
+keyword defined in the same schema object.
+
+However, there is a small set of keywords whose evaluation depend on the
+<i>scope</i> they are in. These keywords are `$ref`, `$dynamicRef`,
+`unevaluatedItems`, and `unevaluatedProperties`. Additionally, there is also a
+set of keywords that <i>affect</i> the scope they are declared in. These keywords
+are `$id`, `$schema`, `$anchor`, and `$dynamicAnchor`.
+
+JSON Schema introduces two types of scopes: the <i>lexical scope</i> and the
+<i>dynamic scope</i>. Understanding how these scopes work is essential for mastering
+some of the most advanced (and often confusing!) features of JSON Schema, such
+as dynamic referencing.
+
+Lexical Scope
+-------------
+
+JSON Schema is a circular data structure. What we mean by this is that as a
+schema author, you may use <i>applicator</i> keywords that take other schemas as
+arguments, effectively creating a tree of schemas. 
+
+Consider the following example. In the left, a JSON Schema consisting of 7
+subschemas (declared using the `properties` and `prefixItems` applicators). In
+the right, a tree representation of the schema, where each subschema
+corresponds to a node in the tree:
+
+![Thinking of a JSON Schema as a tree structure](/img/posts/2024/understanding-lexical-dynamic-scopes/tree.webp)
+
+Thinking of a schema as a tree structure greatly helps understanding lexical
+scopes. Under this analogy, the lexical scope of a schema consists of the
+<i>ancestors</i> of such node, including the node itself. In other words, the
+lexical scope of a schema consists of its parent schemas including the schema
+itself, but not its subschemas.
```

Comment:
I'm not sure that this is a sufficient definition for lexical scope.  In other words, why is lexical scope defined this way?  How does it benefit us to define it this way?

I think it may be helpful to consider a definition for what lexical scope may have been in draft 7, before we needed to distinguish it from dynamic scope.

Reply:
> I think it may be helpful to consider a definition for what lexical scope may have been in draft 7, before we needed to distinguish it from dynamic scope.

Can you link me to this definition? I'm not very familiar with Draft 7, and the specification doesn't seem to mention lexical scopes or even scope in it. Is there a specific section you have in mind?

Reply:
I think that was kinda my point.  Lexical scope definitely existed before we added dynamic scope, but we didn't have a name for it until we needed to distinguish between the two ideas.

This may be something that we need more discussion around, but if others are happy with this for the purpose of this blog post, I'm okay with it.

Probably still something to consider adding to the spec and to the glossary.

Reply:
From @jdesrosiers's [comment](https://github.com/json-schema-org/website/pull/274#pullrequestreview-1837060119) below:

> The lexical scope is just a subset of the dynamic scope that includes only the locations within the current schema resource.

I was also thinking the lexical scope was just the local resource, but in reading the rest of the post, I realized that some portions depend on your definition.  Rather than propose this, I thought I'd just ask why you defined it that way.

Reply:
> I thought I'd just ask why you defined it that way.

I'm just trying to write down what I understand so far, so it's good that we have these discussions. If there is a better way to explain the ideas, in a way that I might have missed, then that's great and we should do that.

> The lexical scope is just a subset of the dynamic scope that includes only the locations within the current schema resource.

That's a very interesting observation, but not sure I fully get it. Lexical & dynamic scopes will typically diverge when references are encountered. There would be schemas from the schema resource that are not in the dynamic scope but would be on the lexical scope, so lexical scope wouldn't always be a subset of the dynamic scope, correct?

Reply:
Maybe "the lexical scope is the current schema resource" is a better way to put it.  @jdesrosiers , thoughts?

Reply:
> There would be schemas from the schema resource that are not in the dynamic scope but would be on the lexical scope

This is not correct. The lexical and dynamic scope are constructed in exactly the same way. There's no way the lexical scope can differ from the dynamic scope other than the resource boundary restrictions.

Reply:
> Maybe "the lexical scope is the current schema resource" is a better way to put it. @jdesrosiers , thoughts?

It _should_ be that simple, but I don't think that's technically the way dynamic scope is defined. If I was writing an article about dynamic references, that's what I would say as a little white lie because that's a far more reasonable way to explain it in the context of dynamic references and dynamic references are the only use-case for dynamic scope. Since this article is about dynamic scope, not dynamic references, I probably wouldn't go with that description here.

Reply:
But what's the point of defining scope outside of the context of references?  Scope exists solely to define how references are resolved.  It seems, then, that scope needs to be defined with that in mind.

If this were my task, I would define:

- Lexical scope is the resource that contains the subschema/keyword being evaluated.  (A fragment-only URI points to the same lexical scope, whereas a non-fragment-only URI indicates a new scope to resolve.)
- Dynamic scope is the ordered set of resources through which evaluation has passed.

These definitions allow you to properly resolve both kinds of references, and I think that is sufficient for the purpose of this post.

Reply:
FWIW, @gregsdennis 's explanation above is also quite simple/elegant. Happy to go with it if we all agree

Reply:
@Relequestual Based on some of your comments, how do you feel with Greg's definition above ^?

Reply:
Maybe one potential minor comment from me:

> Dynamic scope is the ordered set of resources through which evaluation has passed.

If we are saying that the lexical scope if the entire schema resource, should we say that the dynamic scope is the ordered set of "subschemas" instead?

Then, questions around how `$schema` and `$anchor` come into play are resolved by saying that they are within the lexical scope of each of the subschemas in the dynamic scope. Seems to complement it well on my mind

Reply:
> within the lexical scope of each of the subschemas in the dynamic scope

Moving forward, we're removing `$dynamicRef`'s ability to act like a regular `$ref`, meaning that the two referencing mechanics will be completely disjoint.  I think it's easier to understand them this way, and defining one based on the other just confuses things.

Just my opinion, though.

Reply:
Ah, makes sense then. What I'll do is probably leave this PR as-is for a couple more days in case anybody else has any thoughts, then push for your definition above ^, and then address the content accordingly

Reply:
@jviotti @gregsdennis @Relequestual There are so many threads going on, so I'm going to respond to the general themes here. If there's anything I don't cover that you want me to respond to, please ping me.

### Lexical scope and instances
I probably misunderstood Henry and the lexical scope is not related to the evaluation path. Even if I didn't misunderstand Henry, I don't think that interpretation makes sense and I'm more than happy with us all deciding to describe it in a different way. However, in that case, I don't see how the concept of lexical and dynamic scopes converging makes sense. These are two completely different mechanisms doing different things, so how can they ever be the same?

### Dynamic scope: subschemas or resources
I feel that Henry was quite clear that the dynamic scope was intended to be the evaluation path of subschemas, not resources. However, I've always thought that level of granularity was unnecessary and I'd be fine if we all decide that we want to present this concept to people as resources rather than sub-schemas.

At the moment, the only dynamic scope concept we have is dynamic references. That concept only cares about the resource evaluation path because dynamic anchors use a resource-level keyword. I would hope we don't consider adding any more dynamically scoped features, but it's conceivable that a schema-level keyword could could effect dynamically scoped state, in which case the subschema evaluation path becomes relevant. If we want to leave the door open for this kind of thing, we'd want to stick to a subschema evaluation path definition.

### Thoughts on scopes
I think a mistake we commonly make is thinking that some behavior being dependent on the instance is a marker of dynamic scope. For example, the ["multiple dynamic paths"](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L391-L475) test doesn't actually demonstrate dynamic scope. It only shows the need for global state. However, the ["after leaving a dynamic scope, it is not used by a $dynamicRef"](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L478-L515) test does demonstrate dynamic scope as the value of the that global state changes depending on the evaluation path.

### Scope levels
In thinking about scopes, I found it helpful to write out something schema like in a (fake) language that looks more like a programming language where we're more comfortable thinking about scopes. I think this can help demonstrate what I meant about resource-level keywords.

```
// A dynamically scoped global variable that will track the target of the "thingy" dynamic anchor
global thingy

class Main {
  var $id = "https://test.json-schema.org/dynamic-ref-leaving-dynamic-scope/main"
  var $schema = "https://json-schema.org/draft/2020-12/schema"

  main(instance) {
    if (firstScope(instance)) {
      secondScope(instance)
    }
  }
}

class FirstScope {
  var $id = "https://test.json-schema.org/dynamic-ref-leaving-dynamic-scope/first-scope"
  var $schema = "https://json-schema.org/draft/2020-12/schema"

  // Set the value the global "thingy" will have from this point on in the evaluation.
  // This value will be forgotten when evaluation leaves this class
  local thingy = thingy || this.main

  main(instance) {
    type = "number"
  }
}

class SecondScope {
  var $id = "https://test.json-schema.org/dynamic-ref-leaving-dynamic-scope/second-scope"
  var $schema = "https://json-schema.org/draft/2020-12/schema"
  local thingy = thingy || this.main

  main(instance) {
    start(instance)
  }

  isNull(instance) {
    type = "null"
  }
}

class Start {
  var $id = "https://test.json-schema.org/dynamic-ref-leaving-dynamic-scope/start"
  var $schema = "https://json-schema.org/draft/2020-12/schema"

  main(instance) {
    innerScope.anchor(instance)
  }
}

class InnerScope {
  var $id = "https://test.json-schema.org/dynamic-ref-leaving-dynamic-scope/start"
  var $schema = "https://json-schema.org/draft/2020-12/schema"
  local thingy = thingy || this.main

  main(instance) {
    type = "string"
  }

  anchor(instance) {
    thingy(instance)
  }
}
```

The class is like a schema resource and the methods are like subschemas. A class has instance variables are (lexically) in scope for anything in within the class. In a schema, we have a few keywords that behave like this. I call them resource-level keywords. Most keywords would only be visible within their local subschema (or method in this analogy). I call these schema-level keywords.

In this example, the `local` keyword is declaring a dynamically scoped global variable. This is used to set the dynamic anchor. The "thingy" anchor is set at the class level, because dynamic anchors are resource-level. From this example, it might be more clear how it could make sense for someone to define a schema-level keyword that uses dynamic scope. In that case `local` would be used in the method-level rather than the instance-level.

Summary
* Resource-level lexical scope: `$id`, `$schema`, `$anchor`
* Schema-level lexical scope: `type`, `properties`, `if`, `then`, etc.
* Resource-level dynamic scope: `$dynamicAnchor`

Reply:
I think a mistake we commonly make is thinking that some behavior being dependent on the instance is a marker of dynamic scope. For example, the ["multiple dynamic paths"](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L391-L475) test doesn't actually demonstrate dynamic scope. It only shows the need for global state. However, the ["after leaving a dynamic scope, it is not used by a $dynamicRef"](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/dynamicRef.json#L478-L515) test does demonstrate dynamic scope as the value of the that global state changes depending on the evaluation path.

This is a very good example: the test I had the single most difficulty with implementing was the "after leaving a dynamic scope, it is not used by a $dynamicRef" one - because it was the only one that meant I really had to understand the mechanism in play.

Reply:
@jdesrosiers I like your points and the programming-language analogy example. I think that's a good mental model for how scoping should work. I particularly like the *schema resource to class* analogy, the *subschema to method* analogy, and the *resource level lexical scope to member* analogy.

> - Resource-level lexical scope: $id, $schema, $anchor
> - Schema-level lexical scope: type, properties, if, then, etc.
> - Resource-level dynamic scope: $dynamicAnchor

I liked what was suggested at some point, which is that both lexical & dynamic scope operate at the subschema level instead of at the resource level (which feels natural too) and that there are some keywords, like `$anchor`, which conceptually affect every lexical scope of the schema resource they are defined in.